File: updated_code/code/java/1370-increasing-decreasing-string.java
import java.util.Arrays;
import java.lang.System;

public class Solution {
    public String sortString(String s) {
        // Comment added as required by PMD
        // The problem statement is to reorder the string using the given algorithm
        // The code below sorts the string first and then reorders it according to the problem statement's rules
        char[] arr = s.toCharArray();
        Arrays.sort(arr);

        StringBuilder result = new StringBuilder();
        int left = 0, right = arr.length - 1;

        // The while loop iterates until all characters are added to the result
        while (left <= right) {
            // If the left index is less than or equal to the right index, we can pick the smallest character
            if (left <= right) {
                result.append(arr[left]);
                left++;
            }

            // If the left index is less than or equal to the right index, we can pick the smallest character greater than the last appended character
            if (left <= right) {
                result.append(arr[right]);
                right--;
            }
        }

        // Return the result
        return result.toString();
    }
}

File: updated_code/code/python/1370-increasing-decreasing-string.py
Python
from collections import Counter

class Solution:
    def sort_string(self):
        """
        Reorder the string using the following algorithm:

        1. Pick the **smallest** character from `s` and **append** it to the result.
        2. Pick the **smallest** character from `s` which is greater than the last appended character to the result and **append** it.
        3. Repeat step 2 until you cannot pick more characters.
        4. Pick the **largest** character from `s` and **append** it to the result.
        5. Pick the **largest** character from `s` which is smaller than the last appended character to the result and **append** it.
        6. Repeat step 5 until you cannot pick more characters.
        7. Repeat the steps from 1 to 6 until you pick all characters from `s`.

        Args:
        s (str): Input string.

        Returns:
        str: The result string after sorting `s` with this algorithm.
        """
        cnt = Counter(self.s)
        res = []
        while cnt:
            # sort small characters
            small_chars = sorted([c for c in cnt if cnt[c] > 0])
            while small_chars and cnt[small_chars[0]] > 0:
                cnt[small_chars[0]] -= 1
                res.append(small_chars.pop(0))

            # sort large characters
            large_chars = sorted([c for c in cnt if cnt[c] > 0], reverse=True)
            while large_chars and cnt[large_chars[0]] > 0:
                cnt[large_chars[0]] -= 1
                res.append(large_chars.pop(0))

        return ''.join(res)

    def __init__(self, s: str):
        self.s = s

    def sort_string(self) -> str:
        return self.sort_string()